### ISSUE TITLE
Auto-generate main.zng

### ISSUE THREAD

**Issue by kraktus** [HKalbasi/zngur#5]
Hello, thanks for this promising project.

I’ve a rust crate https://github.com/kraktus/rs-retroboard-chess that has sparked the interest of cpp users. I’d like to make it available.

The good thing is that it mostly returns simple types or wrapper around it, however the api surface is still quite high to make a main.zng file manually. Is there a way to generate it from the rust code? Even imperfectly, as a way to save time.

**Comment by HKalbasi** [comment]
I added a very low effort implementation of a `main.zng` generator, you can use it by generating your project's json docs by `cargo +nightly rustdoc -Zunstable-options --output-format=json` and copy the result into `zngur/zngur-autozng/doc.json` and then run autozng using `cargo run`. I did it for your project, and here is the result:
```
type crate::MoveKind {
    #heap_allocated;
    fn new(Option, Option) -> Result;
    fn is_uncapture(self) -> bool;
    fn is_en_passant(self) -> bool;
    fn is_unpromotion(self) -> bool;
    fn to_retro_uci(self) -> String;
}
type crate::UnMove {
    #heap_allocated;
    fn from_retro_uci(&str) -> Result;
    fn new(Square, Square, MoveKind) -> Self;
    fn to_retro_uci(self) -> String;
    fn is_uncapture(self) -> bool;
    fn uncapture(self) -> Option;
    fn is_unpromotion(self) -> bool;
    fn is_en_passant(self) -> bool;
    fn uncapture_square(self) -> Option;
    fn mirror(self) -> Self;
}
type crate::RetroPockets {
    #heap_allocated;
    fn color(self, Color) -> &RetroPocket;
    fn color_mut(self, Color) -> &mut RetroPocket;
    fn from_str(&str, &str) -> Result;
}
type crate::RetroBoard {
    #heap_allocated;
    fn new_no_pockets(&str) -> Result;
    fn new(&str, &str, &str) -> Result;
    fn push(self, &UnMove);
    fn pseudo_legal_unmoves(self, &mut UnMoveList);
    fn legal_unmoves(self) -> UnMoveList;
    fn board(self) -> &Board;
    fn retro_turn(self) -> Color;
    fn us(self) -> Bitboard;
    fn our(self, Role) -> Bitboard;
    fn them(self) -> Bitboard;
    fn their(self, Role) -> Bitboard;
    fn king_of(self, Color) -> Square;
    fn flip_vertical(self);
    fn flip_horizontal(self);
    fn flip_diagonal(self);
    fn flip_anti_diagonal(self);
    fn rotate_90(self);
    fn rotate_180(self);
    fn rotate_270(self);
}
type crate::ParseRetroPocketError {
    #heap_allocated;
}
type crate::RetroPocket {
    #heap_allocated;
    fn decr(self, Role);
}
type crate::ParseRetroUciError {
    #heap_allocated;
}
```
It is broken in multiple ways, and you can use it just as a template. You need to fix these things for it to become usable:
* Names should be full qualified, like `crate::UnMove`, `::shakmaty::Role` or `::std::string::String`.
* Generics are ignored and should be added
* Generic types used in function arguments and return types should be added as separate types, with appropriate methods like `unwrap`.
* `self` types are wrong (I can fix this easily if it is annoying).
* You probably want to do this once for `shakmaty` crate as well.

The `autozng` is a simple 200 line script which you can easily read and improve, and ideally we want to have a full featured `autozng` and it is in the scope of this project, however I currently don't have time to make it complete and polished, but I can add some more features if you desire.

Please tell me if you hit any problems in using zngur, or have some ideas. My own project that triggered me to create zngur became fully Rust, so now there is no real project using it. But I still believe in zngur idea and would really like to see a real project use zngur, so I will help you if you hit any zngur related problem.

**Comment by kraktus** [comment]
Thanks! I actually did something similar 2 years ago: https://github.com/kraktus/cargo-extern-fn and got quite far using proc macros, but it was based on cxx which turned out to be too limited in terms of fondamental type support (not even Option<T>), and seemingly no active development from the maintainers to add them or accept PRs.

Will definitely check it out once I've more time.

**Comment by camio** [comment]
I think it's important to note that an autozng will always be incomplete. It's impossible to know, without reading the C++ code, which of the infinite variations of tuples, for example, are needed on the C++ side.

**Comment by HKalbasi** [comment]
Yes, there never will be a full automatic autozng. The most complete autozng possible could receive a list of end types, for example `Vec<i32>`, `(Foo, &[Bar])`, and others that are directly used in C++, and automatically generate all of their methods and types that are needed for those method (in this example, they are `i32`, `Option<i32>`, `Foo`, `Bar`, `&[Bar]`, ...). But even this can lead to bloat in the generated code and compile times so I think even more manual control is needed at the end.

**Comment by Ryex** [comment]
I've been letting my brain simmer on this one so the last few months and I think I've reached an approach I'm willing to implement and PR.

from the beginning there a series of issues to work out

- [x] Any autozng impl would be incomplete
  - Mostly Solved: `import` statements let you generate code for the public rust side and extend it with cpp side or spesfic generic instances.
  - generation my make it desirable to not require a layout in every type decl. make it a generation time error for it to be missing.

- [ ] layout choice.
  - Proposal: let `autozng` side load a 'size' databases (.json) generated/updated from `-Z print-type-sizes` (like the work in #58) or by hand that maps target and opt level  configuration to size and alignment for fully qualified concrete types. configurably fallback to heap allocation (or a future conservative overallocation option) or error if it tries to generate a type it can't find. database could even include size and offset for fields.

- [ ] Generics: how to we emit the correct concrete types for generics?
  Proposal: type templates and impl declarations.
Templates can be expanded for full concrete types that appear in impl declarations and possibly for types that would otherwise be missing. Autozng could output templates for generic types and functions. impl declarations would then be left to the main spec.

- [ ] Where to get type info for generation. This could be done two ways
  - `rustdoc`: Requires nightly and you'd need to put the zngur generated module / code relent on it behind a feature to hide it from rustdoc  but you get the full benifits of the compiler including conditional compilation and macro expansion.
  - syn parse module files: not as heavy as a compiler call. can be done without nightly. selective of which files to examine. but no macros or conditional compilation.
  - could support both with seperate commands. extra work.

this work would make code gen a 2-3 step process 

- Aquire type size info and choose fallback.
  - Generate or write it and commit it along side the code
  - OR Have a nightly compiler in the compilation env and generate it ever time

- Run autozng to generate any interface zng files. import them into the main spec.

- Run zng cli or build script to generate header and module

**Comment by Ryex** [comment]
I had a thought. 

if autozng uses `rustc_public` and saves the layout of the types in use. we have a zng file recording the types and interface there of we want and presumably a database of different layouts for those types from running autozng for different targets.

a second utility like `zngpatch` or similar could use the data base and "retarget" a zngur spec to use layout information for a new target.

Would this be a viable path do you think?

or perhaps `zngpatch` is the utility that can read compiler errors for the correct size and offset?

if any combination of these ideas speaks to you let me know.
the rubber duck stage is important.

**Comment by Ryex** [comment]
I actually did a github wide code search for `rustc_public` to see how others handled it, saw kani, and dug into what they were doing.

turns out that while not requiring compiling from source it *does* require installation of rustup and the pinned nightly. their bundle include a toolchain file that is just the pinned nightly channel and then when you run `kani setup` it ensures rustup is installed and uses it to install that toolchain to make `kani-compiler` usable.

No other project I could find even gets that far with usability, most not having releases and usually encouraging rustup and `cargo install` to pickup the toolchain requirements and compile from source.

**Comment by HKalbasi** [comment]
@Ryex did you receive the invite?
